tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Session.java

```diff
@@ -269,13 +312,11 @@ public Runner addTarget(String operation) {
      * @param operation the operation to execute
      * @return this session runner
      * @throws IllegalArgumentException if the operation is not a {@link GraphOperation}
+     * @throws IllegalStateException if the operation is not from the session's graph.
      */
     public Runner addTarget(Operation operation) {
-      if (!(operation instanceof GraphOperation)) {
-        throw new IllegalArgumentException(
-            "Operation of type "
-                + operation.getClass().getName()
-                + " is not supported in graph sessions");
+      if (operation.env() != graph) {
+        throw new IllegalStateException("Can't fetch operation " + operation + ", it is from a different graph.");
```

Comment:
This exception message is a little different, and could be confusing when debugging if you pass in an eager op somehow as it won't come from a graph and the message might lead you up the wrong path. Maybe "it is either an eager operation or from a different graph"? Similarly for the other places with this exception message that accept an `Operation`.